python/huggingfaceserver/huggingfaceserver/model.py

```diff
@@ -357,35 +299,230 @@ def postprocess(
         elif (
-            self.task == MLTask.text_generation.value
-            or self.task == MLTask.text2text_generation.value
+            self.task == MLTask.text_generation
+            or self.task == MLTask.text2text_generation
         ):
             outputs = self.tokenizer.batch_decode(outputs, skip_special_tokens=True)
             return get_predict_response(request, outputs, self.name)
         else:
             raise ValueError(
                 f"Unsupported task {self.task}. Please check the supported `task` option."
             )
+
+    def _handle_request(self, req: _GenerateRequest):
+        """
+        Handle a single generation request
+        """
+
+        response_queue, kwargs, params = (
+            req["response_queue"],
+            req["kwargs"],
+            req["params"],
+        )
+
+        def queue_put(outputs):
+            self._loop.call_soon_threadsafe(response_queue.put_nowait, outputs)
+
+        if "seed" in params and params["seed"] is not None:
+            set_seed(params["seed"])
+
+        if params["stream"]:
+            streamer = TextIteratorStreamer(
+                self.tokenizer, skip_prompt=not params.get("echo", False)
+            )
+            thread = Thread(
+                target=self.model.generate, kwargs={**kwargs, "streamer": streamer}
+            )
+            thread.start()
+            # Consume the tokens one by one and add them to the queue
+            for output in streamer:
+                if output != "":
+                    queue_put(output)
+            # Put None to indicate we are finished
+            queue_put(None)
+        else:
+            output_start = (
+                0 if params.get("echo", False) else kwargs["input_ids"].shape[-1]
+            )
+            outputs = self.model.generate(**kwargs)
+            outputs = self.tokenizer.batch_decode(
+                outputs[:, output_start:], skip_special_tokens=True
+            )
+            queue_put(outputs)
+
+    @torch.no_grad()
+    def _process_requests(self):
+        """
+        Process requests from the request queue in a background thread.
+        This ensures we don't block the event loop while running generation.
+        """
+        while True:
+            req = self._request_queue.get()
+
+            # If request is None we should stop processing
+            if not req:
+                break
+
+            self._handle_request(req)
+
+    def _submit_request(
+        self, kwargs: Dict[str, Any], params: CompletionCreateParams
+    ) -> asyncio.Queue:
+        """
+        Add a request to the request queue to be processed. Results for this request
+        will be pushed to the returned async queue.
+        """
+        if not hasattr(self, "_loop"):
+            self._loop = asyncio.get_running_loop()
+        req = _GenerateRequest(
+            kwargs=kwargs, params=params, response_queue=asyncio.Queue()
```

Comment:
So we create a queue per request?

Reply:
Correct. This is adapted from the [example in the Huggingface Transformers docs](https://huggingface.co/docs/transformers/pipeline_webserver#using-pipelines-for-a-webserver).